### DISCUSSION TITLE
Health watcher majority formula

### DISCUSSION THREAD

**Discussion by lynshi** [microsoft/CCF#3817] [General] [answered]
The [`health_watcher.py` script](https://github.com/microsoft/CCF/blob/a16d81be9680e6dd950c37381a4e264bff35d857/tests/infra/health_watcher.py#L55) uses the following logic to determine the majority:
```python
majority = (len(nodes) + 1) // 2
```

This logic isn't accurate for networks with an even number of nodes, right (disregarding how common such network configurations are)? `2f + 1` nodes are required to tolerate `f` failures, so a 5 node network is required to tolerate 2 failures. Then, `majority` in a 4 node network should be `3` as only 1 failure is tolerated, but this formula gives `majority = (4 + 1) // 2 = 2`.

If I'm not mistaken, `majority = math.ceil((len(nodes) + 1) / 2)` is more accurate.

**Comment by jumaffre** [comment, marked as answer]
Well spotted! As `heath_watcher.py` mostly is a sample script, we don't test in a wide range of scenarios and network configuration, so thank you very much for raising this.

I've fixed this in https://github.com/microsoft/CCF/pull/3820.
